### ISSUE TITLE
[linkage = "weak"] doesn't work for const fn

### ISSUE THREAD

**Issue by HaoboGu** [rust-lang/rust#134451]
I'm trying to define a "weak" const in a library, which allows users to redefine their const. On stable Rust, there's no way to do it. On nightly, `linkage` feature seems the solution, but I found that it doesn't work as well. The following is the repro code:

```rust
// In library code:
#![feature(linkage)]

#[linkage = "weak"]
#[no_mangle]
const fn get_foo() -> i32 {
    0
}

pub const FOO: i32 = get_foo();

pub fn hello() {
    println!("hello: {}, {}", FOO, get_foo());
}

// In application code:
use linkage_weak_bug::hello;

#[no_mangle]
const fn get_foo() -> i32 {
    1
}

fn main() {
    // Output: hello: 0, 1
    // Expected: hello: 1, 1
    hello();
}

```

**Comment by HaoboGu** [comment]
I think it's valuable to have "weak constants" in Rust, just disallowing weak linkage for const fns is not we want.

For example, in no_std env, it's very common to define a global channel with a fixed size:

```rust
// This is a simplified code
const CH_SIZE: usize = 4;
static CHANNEL: Channel<CH_SIZE> = Channel::new();
```

And, as a library author, we may want the `CH_SIZE` to be customizable by library users. It's impossible for Rust without using weak linkage now.

**Comment by bjorn3** [comment]
`CHANNEL` gets fully codegened when compiling the crate that contains it. To do this, it's full type needs to be known, which in turn requires `CH_SIZE` to be known. Note that the very fact if compilation succeeds or not and which functions get codegened can depend on the value of a const item or const function. As such weak linkage for const items and const functions can't work. And that is not even taking into account the fact that const items by definition don't have a symbol name unlike statics, but weak linkage matches on symbol name.

**Comment by workingjubilee** [comment]
Indeed, we may wish to investigate making this a compiler error.

But first, while an expert on the linking side of things has spoken, let us consult the experts on the CTFE side of things. cc @rust-lang/wg-const-eval do any of you wish to entertain the possibility of a const fn having its symbol name overwritten at runtime, and users like @HaoboGu consequently seriously misunderstanding the model?

**Comment by workingjubilee** [comment]
@HaoboGu The idea of "generic singletons" is an interesting idea, but it's absolutely not the way things work, and if it did work, it would have to be designed in such a way as to actually be evaluated _without_ taking linkage into account.

**Comment by RalfJung** [comment]
No I don't think `const fn` intends to ever support terrible linker hacks like that. ;) If there's a use-case here, it should be solved with proper, type-safe Rust mechanisms, not at the linker level.

Even on regular `fn` this is rather questionable. What even is the semantics here? When you call `get_foo` is non-deterministically calls *some* function with that name? AFAIK currently we have a safety requirement on `no_mangle` that there must be only one function with that name, so technically your code has UB. We probably want to djust this to support weak linkage since unfortunately people want to use their terrible linker hacks, but we should do so carefully.

**Comment by workingjubilee** [comment]
Getting slightly further afield, but `"weak"` isn't even a coherent runtime concept on all OS, and if memory serves the way it is handled for MachO and ELF can be slightly, subtly different.

**Comment by HaoboGu** [comment]
Thanks for everyone's explanation! Now I know doing it when linking is not proper :)

> If there's a use-case here, it should be solved with proper, type-safe Rust mechanisms

I agree! Currently it's very difficult to achieve this, especially in embedded world. There are many ways to "bypass" it, for example, [embassy](https://github.com/embassy-rs/embassy/blob/main/embassy-time/Cargo.toml#L61-L381) generates lots of feature gates, let users choose the tich-hz via feature gate, and [toml-cfg](https://github.com/jamesmunns/toml-cfg/) does a hacky way via proc-macro, analyzing the out dir of all dependencies.

> We have mechanisms for that already (associated consts in a trait).

It's a little bit weird if we convert a config struct to a trait, and tell users to fill an associated const if they want to customize the config. But yeah, I see this should work. Thanks for the pointer!

**Comment by RalfJung** [comment]
> The idea of "generic singletons" is an interesting idea, but it's absolutely not the way things work, and if it did work, it would have to be designed in such a way as to actually be evaluated _without_ taking linkage into account.

Speaking of singletons, this reminds me of https://github.com/rust-lang/rfcs/pull/3632 / https://github.com/rust-lang/rfcs/pull/3635. So maybe there is a space for "externally defineable consts" in a similar vein?

**Comment by workingjubilee** [comment]
Plausible. cc @m-ou-se if you have any thoughts (or not if not, it's just a data point of possible interest).

**Comment by m-ou-se** [comment]
Yeah that's definitely on our list. We're now calling our project "externally implementable _items_", and consts are also on our list of items, although not our first priority. We'll see how far we can get. :)

**Comment by workingjubilee** [comment]
Cool!

This issue is now about diagnostics! We should at least warn, probably compile error on this.
